storage/replica_raftstorage.go

```diff
@@ -473,8 +469,13 @@ func (r *Replica) applySnapshot(snap raftpb.Snapshot) error {
 
 	rangeID := r.RangeID
 
-	// First, save the HardState.  The HardState must not be changed
-	// because it may record a previous vote cast by this node.
+	// First, save the HardState. The HardState must not be changed
+	// because it may record a previous vote cast by this node. This is
+	// usually unnecessary because a snapshot is nearly always
+	// accompanied by a new HardState which incorporates both our former
+	// state and new information from the leader, but in the event that
+	// the HardState has not changed, we want to use our own previous
+	// HardState and not one that was transmitted via the snapshot.
 	hardStateKey := keys.RaftHardStateKey(rangeID)
 	hardState, _, err := engine.MVCCGet(r.store.Engine(), hardStateKey, roachpb.ZeroTimestamp, true /* consistent */, nil)
```

Comment:
are you consciously not loading from the batch? It seems that the engine can't (or at least shouldn't) have anything newer.